keps/sig-apps/3973-consider-terminating-pods-deployment/README.md

```diff
@@ -256,6 +268,93 @@ To satisfy the requirement for tracking terminating pods, and for implementation
+| T1   | any amount       | 50           | 30           | 20           | 100          | 100                       | 110                                  | -           |
+| T2   | any amount       | 61           | 35           | 24           | 120          | 120                       | 130                                  | 1.182       |
+| T3   | any amount       | 66           | 38           | 26           | 130          | 130                       | 140                                  | 1.077       | 
+
+With the `TerminationComplete` PodReplacementPolicy, scaling cannot proceed immediately if there
+are terminating pods present, in order to adhere to the Deployment constraints. We need to scale
+some ReplicaSets fully and some partially. And we have to postpone scaling to the future when
+terminating pods disappear.
+
+A single scale operation occurs at T2.
+
+| Time | Terminating Pods | RS1 Replicas | RS2 Replicas | RS3 Replicas | All RS Total | Deployment .spec.replicas | Deployment .spec.replicas + MaxSurge | Scale ratio |
+|------|------------------|--------------|--------------|--------------|--------------|---------------------------|--------------------------------------|-------------| 
+| T1   | 15               | 50           | 30           | 20           | 100          | 100                       | 110                                  | -           |
+| T2   | 15               | 61           | 34           | 20           | 115          | 120                       | 130                                  | 1.182       |
+| T3   | 5                | 61           | 35           | 24           | 120          | 120                       | 130                                  | -           | 
+| T4   | 0                | 61           | 35           | 24           | 120          | 120                       | 130                                  | -           | 
+
+To proceed with the scaling in the future (T3), we need to remember both `replicasBeforeScale` and
+`deploymentMaxReplicasBeforeScale` to calculate the original scale ratio. The terminating pods can
+take a long time to terminate and there can be many steps and ReplicaSet updates between T2 and T3.
+If we were tp use the current number of ReplicaSet or Deployment replicas in any of these steps
+(including T3), we would calculate an incorrect scale ration.
+
+- `deploymentMaxReplicasBeforeScale` is already stored in the `deployment.kubernetes.io/max-replicas`
+  ReplicaSet annotation. The main change is that we need to keep the old Deployment max replicas
+  value in the annotation until the partial scale for a ReplicaSet is complete.
+- To remember `replicasBeforeScale`, we will introduce a new annotation called
+  `deployment.kubernetes.io/replicaset-replicas-before-scale`, which will be added to the
+  Deployment's ReplicaSets that are being partially scaled. This annotation will be removed once
+  the partial scaling is complete.
```

Comment:
What adds the new annotation? Is it the control plane or kubectl? Let's specify that.

Reply:
Control plane - deployment controller. Updated